infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -0,0 +1,342 @@
+                            }
+
+                            return stripe;
+
+                        })
+                        .collect(Collectors.toList());
+
+                return new LayoutSegment(
+                        layoutSegment.getReplicationMode(), layoutSegment.getStart(),
+                        layoutSegment.getEnd(), newStripes);
+            } else {
+                return new LayoutSegment(
+                        layoutSegment.getReplicationMode(), layoutSegment.getStart(),
+                        layoutSegment.getEnd(), layoutSegment.getStripes());
+            }
+        }).collect(Collectors.toList());
+        Layout newLayout = new Layout(layout);
+        LayoutBuilder builder = new LayoutBuilder(newLayout);
+        return builder.setSegments(segments).build();
+    }
+
+    /**
+     * Updates the initial layout given the restored transferred segments.
+     *
+     * @param segments   A list of restored transferred segments.
+     * @param initLayout Initial layout.
+     * @return A new, updated layout.
+     */
+    public Layout updateLayoutAfterRedundancyRestoration(
+            List<CurrentTransferSegment> segments, Layout initLayout) {
+        return segments
+                .stream()
+                .reduce(initLayout,
+                        (layout, segment) -> restoreRedundancyForSegment(segment, layout),
+                        (oldLayout, newLayout) -> newLayout);
+    }
+
+    /**
+     * Returns true if a given layout requires redundancy restoration.
+     *
+     * @param layout Current layout.
+     * @param server Current server.
+     * @return True, if requires.
+     */
+    public static boolean requiresRedundancyRestoration(Layout layout, String server) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+            return layout.getSegments().stream().anyMatch(segment -> !segment.getAllLogServers()
+                    .contains(server));
+        }
+    }
+
+    /**
+     * Returns true if after adding a server to the first segment, the segments can be merged.
+     *
+     * @param layout Current layout.
+     * @param server The current node.
+     * @return True is there is a redundancy restoration is needed.
+     */
+    public static boolean requiresMerge(Layout layout, String server) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+            int firstSegmentIndex = 0;
+            int secondSegmentIndex = 1;
+            Layout copy = new Layout(layout);
+            LayoutStripe firstStripe = copy.getFirstSegment().getFirstStripe();
+            firstStripe.getLogServers().add(server);
+            return Sets.difference(
+                    copy.getSegments().get(secondSegmentIndex).getAllLogServers(),
+                    copy.getSegments().get(firstSegmentIndex).getAllLogServers()).isEmpty();
+        }
+    }
```

Comment:
In continuing to our discussion.

> Every time I leave a comment say there is an issue, I did code research and thought about it before doing it. 

Well, probably you need to make code research more carefully.
![image](https://user-images.githubusercontent.com/314292/68067702-7cfc8380-fd08-11e9-963c-95cdf6669191.png)

Also, Javadoc clearly explains the logic: `Returns true if after adding a server to the first segment, the segments can be merged.`
Which means you are not correct in your comment.

P.S.
What were those critical issues in the code that you had to mention it clearly in the final comment?
To me, there could be a really critical issue in the business logic that breaks the app to deserve to be mentioned in the final comment that the code must not go because of the issue.

Don't get me wrong, we all make mistakes like in this PR Pavel forgot to assign a variable in ServerConfigurator class and you found that small mistake, that's great, that's why we are doing reviews (btw, those mistakes won't break anything in the current version).
But is it worth it to mention in the final comment, and expose it as a very critical problem?

Imagine we all will leave such comments in any PR-s, when any one of us finds any small mistake in the code or just believe that the mistake is there?

To be clear, we are not talking about this PR specifically, we discuss in general, and it happened many times before (offline) if someone believes that some code is broken or there is a critical problem in that piece of code someone must be able to prove his statements but not just randomly generate statements about code or functionality just because he believes that there is the problem when he didn't have good context about the code.

Reply:
@xnull Well I guess you haven't got my example here. Let me explain to you on Monday if you would like. I want to put very clearly here that the final comment is NOT oppose or being harsh on this patch, just telling Pavel that I found potential issues in his PR and he should take a look and fix if it's true and meanwhile I will continue reviewing. For me, "found some issues" in the final comment is not a harsh phrase. This is very normal sentence and I don't know why you see it as a strongly negative expression, maybe my english is not good enough to let you feel like this.  As for the "request change", this is simply to prevent other people approving and it got merged immediately without discussion and consensus on some issues. Because you already approved and one more approval could result in this patch checked in immediately, but this is a large patch and need some time to soak and discuss.  Anyway If you have any questions, come to me and I'm more than welcome to discuss. This is Pavel's pull request and let's NOT spam it arguing the non-tech stuffs.

Reply:
In the scenario that you described, C is in the last segment, which means it is either performing heal at the moment, so it won't be executing this workflow or tried to perform restore redundancy during a heal but failed.
If C is not executing this workflow and finishes well with heal we will get 
[A, C], [A, B, C], [A, B, C].
If B is executing heal or B is executing restore redundancy the segments will merge into 1 at the end by B.
If C failed during a heal, 
C can not merge the first two segments since (A, B) - (A, B, C) won't produce an empty set. B can do that because (A, B) - (A, B) would produce an empty set. After the first two segments are merged by B, C can merge the segments because (A, B, C) - (A, B, C) would produce and the empty set and the segments will merge into 1 at the end by C. 

The above case actually won't happen in the first place because of the presense of B in second segment. We actually restore B to each segment of the layout only after all the segments are restored.